Creating new TypeName without reparsing + usage in NRBF

adamsitnik opened the pull request:
@jkotas I've tried to implement the APIs we have discussed in #102263 and so far it looks good.

I need to spend more time testing `MakeGenericTypeName` and thinking about all edge cases.

contributes to #102263
fixes #106022

adamsitnik committed:
new TypeName methods

adamsitnik committed:
do not join the type and library name strings when UndoTruncatedTypeNames is set to false, parse assembly name once and just reuse it

adamsitnik committed:
cache known primitive type names (and SZArrays of these)

adamsitnik committed:
add more tests and finish WithAssemblyName implementation

jkotas commented on src/libraries/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs:
```diff
@@ -2445,6 +2445,12 @@ public sealed partial class TypeName
         public System.Reflection.Metadata.TypeName GetGenericTypeDefinition() { throw null; }
         public System.Reflection.Metadata.TypeName GetElementType() { throw null; }
         public int GetNodeCount() { throw null; }
+        public System.Reflection.Metadata.TypeName MakeArrayTypeName() { throw null; }
```
Do you also plan to add `MakeSimpleTypeName` and `MakeNestedTypeName`?

jkotas replied:
I think it would be better to expose those APIs and build `WithAssemblyName` as a helper in S.F.Nrbf. The behavior of WithAssemblyName is very Nrbf specific. I do not think it is generally usable method.

adamsitnik replied:
> Do you also plan to add `MakeSimpleTypeName` and `MakeNestedTypeName`?

`MakeSimpleTypeName` would need to parse the name anyway. I think it's simpler to just use `TypeName.Parse`. Unless I am missing something (the method would throw for non-simple names?)?

`MakeNestedTypeName` I currently don't see any real-life usage for it and would prefer to wait for customer feedback about it.

adamsitnik replied:
> I think it would be better to expose those APIs and build `WithAssemblyName` as a helper in S.F.Nrbf. 

But I don't know how to do that with the current APIs without the re-parsing.


> The behavior of `WithAssemblyName` is very Nrbf specific. I do not think it is generally usable method.

I've seen a need in WinForms but it's also related to BinaryFormatter:

https://github.com/dotnet/winforms/blob/31176a4d203711e766b8b587414d04c27d8cc173/src/System.Windows.Forms/src/System/Resources/ResXSerializationBinder.cs#L54-L61

I believe that having this method and for example allowing to pass null to clear all assembly names is generally usable.

jkotas replied:
> I don't know how to do that with the current APIs without the re-parsing.

You just need to validate that the type name does not contain invalid escape sequence. It is much cheaper that full reparsing.

> I believe that having this method and for example allowing to pass null to clear all assembly names is generally usable.

I can believe that it is useful in the world of binary serializers that are not the primary design point of these APIs. I do not think it is useful for the world of ECMA and compilers.

adamsitnik replied:
> You just need to validate that the type name does not contain invalid escape sequence. It is much cheaper/simpler than full reparsing.

Now I get the point, moreover I see the value and would use it myself in Nrbf. I've extended the proposal in https://github.com/dotnet/runtime/issues/102263. It seems that https://apireview.net/backlog is pretty packed, I will implement `static TypeName MakeSimpleTypeName` when I am back from my vacations and extend the proposal again.

> I can believe that it is useful in the world of binary serializers that are not the primary design point of these APIs. I do not think it is useful for the world of ECMA-335, compilers and similar tools.

And that API needs to serve both, so I don't see why we should not add it.

jkotas replied:
> And that API needs to serve both, so I don't see why we should not add it.

`WithAssemblyName` is a nice-to-have convenience API for binary serializers. It is not required to make the binary serializers use case work. Binary serializers can have it implemented as a local helper just fine.

jkotas replied:
> MakeNestedTypeName I currently don't see any real-life usage for it and would prefer to wait for customer feedback about it.

It is necessary for completeness so that one can implement arbitrary type name builder or rewriter (`WithAssemblyName` is an example of such rewriter).

adamsitnik commented:
> I need to spend more time testing `MakeGenericTypeName` and thinking about all edge cases.

The `MakeGenericType` name from reflection works only for generic type definitions:

https://github.com/dotnet/runtime/blob/9e57de2b2a71b449057ffe75823d332e44ffe8fa/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/TypeInfos/RuntimeTypeInfo.cs#L431

Should `MakeGenericTypeName` have similar limitation? Otherwise, we are going to allow for creating generic type names where `typeName.IsConstructedGenericType` is pointer, array etc. Such name would be illegal for runtimes we have today.

jkotas commented:
> Should MakeGenericTypeName have similar limitation? 

Sounds good to me. I assume that you are thinking about requiring `IsSimpleName` to be true. Correct?

adamsitnik commented:
> Sounds good to me. I assume that you are thinking about requiring `IsSimpleName` to be true. Correct?

`IsSimple` or `IsConstructedGenericType` (so users could change generic args of given generic type)

jkotas commented:
> IsConstructedGenericType (so users could change generic args of given generic type)

I do not think that regular reflection APIs allow this. Why would we want to allow it? If users want to change generic args of constructed generic type, they should get the generic type definition first on their side.

adamsitnik committed:
finish the implementation

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into extendTypeName

adamsitnik committed:
add missing IntPtr and UIntPtr support (discovered once I've re-enabled BF tests in main and synced the branch)

adamsitnik committed:
use TypeNameMatches in more places

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into extendTypeName

# Conflicts:
#	src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/SerializationRecord.cs

adamsitnik committed:
address API review feedback

adamsitnik committed:
remove CreateSimpleTypeName, introduce MakeSimpleTypeNameThrowsForNonSimpleNames:

- all TypeName instances remain always valid and escaped
- it's impossible to detect all sequences that are escaped

adamsitnik committed:
handle nested names properly

adamsitnik committed:
update the code after reading it again on GH

jkotas reviewed (commented):
I am not familiar enough with the NRBF to review the changes in NRBF. You should ask Jeremy or Levi to review that. I have reviewed S.R.Metadata changes only.

jkotas commented:
> fixes https://github.com/dotnet/runtime/issues/102263

This is missing `CreateSimpleTypeName` API from this proposal.

adamsitnik commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -392,5 +411,99 @@ public int GetArrayRank()
+        /// <param name="assemblyName">Assembly name.</param>
+        /// <returns>Created simple name.</returns>
+        /// <exception cref="InvalidOperationException">The current type name is not simple.</exception>
+        public TypeName MakeSimpleTypeName(AssemblyNameInfo? assemblyName)
```
> All other `Make` method take the `this` instance as a whole and make a new thing from it.

> This method is different. It replaces a property on `this` instance with new value, it does not take the `this` instance as a whole like all other methods.

It does that for a reason: I wanted to hide the complexity of changing nested type names from the user.

Let's condiser `CreateSimpleTypeName` so I can better express what I mean:

```cs
public static TypeName CreateSimpleTypeName(
        string metadataName,
        TypeName? declaringType = null,
        AssemblyNameInfo? assemblyInfo = null);
```

What if the input is following:
- `declaringType`: a type called `MyNamespace1.Declaring`
- `metadataName`: `MyNamespace2.Declaring+Nested`

We either end up with an invalid name (the declaring and nested type have to belong to the same namespace and assembly) or a need to verify them.

There is one more question: is the `metadataName` supposed to be a full name or just the name for nested types?

With the `public TypeName MakeSimpleTypeName(AssemblyNameInfo? assemblyInfo)` the users are allowed to change the `AssemblyName` and don't need to worry about the nested types (it's handled for them). There is no need to re-parse or validate anything and it is exactly what I need.

> The name does not look right to me.
> If we really need to have this method, `WithAssemblyName` that you have proposed previously sounds better to me.

Naming is hard and subjective. For me, the `MakeSimpleTypeName` name suggests that it works only for simple types and can be used to create a new instance with a different assembly name.

I also like the previous proposal (`WithAssemblyName`), but it does not suggest anything that it's going to work only for simple names.

> This method is a small optimization of the BF scenario. (I am not convinced that it is necessary to make NRBF secure.) 

The BF scenario is important to me and I am not convinced that I don't need it.

@jkotas If you believe that this name is wrong I am open to marking it as blocking and re-discussing it tomorrow with the API review board.

adamsitnik replied:
> Do you agree that it is just a minor optimization, and that this exact API is not required to make the Nrbf-reader secure?

I don't agree. It's not a minor optimization, it is my requirement that was pointed out by @GrabYourPitchforks in code review (lack of any kind of re-parsing). Why otherwise would I be pushing for including it in RC1?

jkotas replied:
Can you explain the rationale behind this requirement?

I think that the requirement should be to avoid any non-linear work (where N is the size of the input). For example, allocating a substring (that I expect that many use cases will end up doing) or re-validating whether the string escaped properly should be fine since it is still linear amount of work.

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into extendTypeName

adamsitnik committed:
address code review feedback

JeremyKuhne commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/MemberTypeInfo.cs:
```diff
@@ -142,63 +142,25 @@ internal TypeName GetArrayTypeName(ArrayInfo arrayInfo)
+            BinaryType.Object => TypeNameHelpers.GetPrimitiveTypeName(TypeNameHelpers.ObjectPrimitiveType),
+            BinaryType.ObjectArray => TypeNameHelpers.GetPrimitiveSZArrayTypeName(TypeNameHelpers.ObjectPrimitiveType),
+            BinaryType.SystemClass => (TypeName)additionalInfo!,
+            _ => ((ClassTypeInfo)additionalInfo!).TypeName,
```
We're losing an assert here. Is the assert no longer valid? Is there another consistency check we can add?

adamsitnik replied:
We could re-add the assert here, but I don't see much value in that. The reader is well guarded against unknown types:
https://github.com/dotnet/runtime/blob/6e440de5b435aee32673376773c58652d1474f49/src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/BinaryReaderExtensions.cs#L32-L40

And if there is a type mismatch, the cast is going to throw anyway.

JeremyKuhne commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/TypeNameHelpers.cs:
```diff
@@ -0,0 +1,185 @@
+                ObjectPrimitiveType => "System.Object",
+                IntPtrPrimitiveType => "System.IntPtr",
+                UIntPtrPrimitiveType => "System.UIntPtr",
+                _ => "System.UInt64",
```
Do we need a catch-all, or can we assert here when we see a PrimitiveType we don't expect (beyond just None and Null)?

JeremyKuhne commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/TypeNameHelpers.cs:
```diff
@@ -0,0 +1,185 @@
+        else if (typeof(T) == typeof(UIntPtr)) return UIntPtrPrimitiveType;
+        else
+        {
+            Debug.Assert(typeof(T) == typeof(string));
```
Is it ever valid to return string for other T's? Can we be explicit for the string case and throw instead of asserting?

JeremyKuhne reviewed (commented):
Added a few consistency check questions.

adamsitnik committed:
address code review feedback

bartonjs commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/TypeNameHelpers.cs:
```diff
@@ -0,0 +1,187 @@
+    internal const PrimitiveType ObjectPrimitiveType = (PrimitiveType)19;
+    internal const PrimitiveType IntPtrPrimitiveType = (PrimitiveType)20;
+    internal const PrimitiveType UIntPtrPrimitiveType = (PrimitiveType)21;
+    private static readonly TypeName?[] s_PrimitiveTypeNames = new TypeName?[(int)UIntPtrPrimitiveType + 1];
+    private static readonly TypeName?[] s_PrimitiveSZArrayTypeNames = new TypeName?[(int)UIntPtrPrimitiveType + 1];
+    private static AssemblyNameInfo? s_CoreLibAssemblyName;
```
Incorrect casing in these names

```suggestion
    private static readonly TypeName?[] s_primitiveTypeNames = new TypeName?[(int)UIntPtrPrimitiveType + 1];
    private static readonly TypeName?[] s_primitiveSZArrayTypeNames = new TypeName?[(int)UIntPtrPrimitiveType + 1];
    private static AssemblyNameInfo? s_coreLibAssemblyName;
```

bartonjs commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/TypeNameHelpers.cs:
```diff
@@ -0,0 +1,187 @@
+        TypeName? typeName = s_PrimitiveTypeNames[(int)primitiveType];
+        if (typeName is null)
+        {
+            string fullName = primitiveType switch
+            {
+                PrimitiveType.Boolean => "System.Boolean",
+                PrimitiveType.Byte => "System.Byte",
+                PrimitiveType.SByte => "System.SByte",
+                PrimitiveType.Char => "System.Char",
+                PrimitiveType.Int16 => "System.Int16",
+                PrimitiveType.UInt16 => "System.UInt16",
+                PrimitiveType.Int32 => "System.Int32",
+                PrimitiveType.UInt32 => "System.UInt32",
+                PrimitiveType.Int64 => "System.Int64",
+                PrimitiveType.UInt64 => "System.UInt64",
+                PrimitiveType.Single => "System.Single",
+                PrimitiveType.Double => "System.Double",
+                PrimitiveType.Decimal => "System.Decimal",
+                PrimitiveType.TimeSpan => "System.TimeSpan",
+                PrimitiveType.DateTime => "System.DateTime",
+                PrimitiveType.String => "System.String",
+                ObjectPrimitiveType => "System.Object",
+                IntPtrPrimitiveType => "System.IntPtr",
+                UIntPtrPrimitiveType => "System.UIntPtr",
+                _ => throw new ArgumentOutOfRangeException(paramName: nameof(primitiveType), actualValue: primitiveType, message: null)
```
I'm on the fence whether I like this lazy initialization because it shows the symbolic name to value matching, or if it should just be declared as `private static readonly string?[] s_primitiveTypeNames = { "System.Boolean", "System.Byte", ..., null, null, "System.SomethingElse", ... };`

bartonjs replied:
Oh, right, this then takes the string and does further work (TypeName.Parse) before squirrelling it away in the array.

OK, this is probably the better approach.

bartonjs commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/TypeNameHelpers.cs:
```diff
@@ -0,0 +1,187 @@
+
+    internal static PrimitiveType GetPrimitiveType<T>()
+    {
+        if (typeof(T) == typeof(bool)) return PrimitiveType.Boolean;
+        else if (typeof(T) == typeof(byte)) return PrimitiveType.Byte;
+        else if (typeof(T) == typeof(sbyte)) return PrimitiveType.SByte;
+        else if (typeof(T) == typeof(char)) return PrimitiveType.Char;
+        else if (typeof(T) == typeof(short)) return PrimitiveType.Int16;
+        else if (typeof(T) == typeof(ushort)) return PrimitiveType.UInt16;
+        else if (typeof(T) == typeof(int)) return PrimitiveType.Int32;
+        else if (typeof(T) == typeof(uint)) return PrimitiveType.UInt32;
+        else if (typeof(T) == typeof(long)) return PrimitiveType.Int64;
+        else if (typeof(T) == typeof(ulong)) return PrimitiveType.UInt64;
+        else if (typeof(T) == typeof(float)) return PrimitiveType.Single;
+        else if (typeof(T) == typeof(double)) return PrimitiveType.Double;
+        else if (typeof(T) == typeof(decimal)) return PrimitiveType.Decimal;
+        else if (typeof(T) == typeof(DateTime)) return PrimitiveType.DateTime;
+        else if (typeof(T) == typeof(TimeSpan)) return PrimitiveType.TimeSpan;
+        else if (typeof(T) == typeof(string)) return PrimitiveType.String;
+        else if (typeof(T) == typeof(IntPtr)) return IntPtrPrimitiveType;
+        else if (typeof(T) == typeof(UIntPtr)) return UIntPtrPrimitiveType;
+        else
```
While I personally think that this looks somewhat reasonable, it is prohibited by our style guidelines: https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md, item 18, first bullet ("Never use single-line form")

Consider expression-switch

adamsitnik replied:
@bartonjs I've tried using switch statement and I was not able to get it to compile (the compiler expect a constant value and `typeof(bool)` is not a const for it as of now).

Using braces turns this 22 lines into 77 lines and makes in unreadable.

According to https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md:

> Braces may be omitted only if the body of every block associated with an if/else if/.../else compound statement is placed on a single line.

So I've changed the last `else` block to be also a single line:

```diff
-else
-{
-    Debug.Fail($"{typeof(T).Name} was not expected.");
-    throw new InvalidOperationException();
-}
+else throw new InvalidOperationException();
```

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -392,5 +411,99 @@ public int GetArrayRank()
+                ? DeclaringType.MakeSimpleTypeName(assemblyName)
+                : null;
+
+            return new TypeName(fullName: FullName,
```
`FullName` includes both the declared type name and the nested type name. It means that this creates non-linear amount of work for types nested multiple times (e.g. try it on something like `A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A`). That does not sound right. Should this be just the name for nested types?

jkotas replied:
Also, is the parser missing a node count check for nested types? I would expect `A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A` to fail to parse with default  `TypeNameParseOptions.MaxNodes`.

adamsitnik replied:
> FullName includes both the declared type name and the nested type name. It means that this creates non-linear amount of work for types nested multiple times

Great catch! I was able to solve it by passing `_fullName` and `_nestedNameLength` to the ctor. It's going to help us avoid any allocations (and keep the promise this API makes ;) )

https://github.com/dotnet/runtime/blob/b080dc2490dfd238d3671e4721308aef39ef14c3/src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs#L148-L153

adamsitnik replied:
> Also, is the parser missing a node count check for nested types? 

That is true, but it's a separate bug. Will send PR shortly.

bartonjs commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/TypeNameHelpers.cs:
```diff
@@ -0,0 +1,187 @@
+        return typeName;
+    }
+
+    private static ArraySegment<char> GetAssemblyQualifiedName(string typeName, string libraryName)
```
Consider renaming to `RentAssemblyQualifiedName`, so it's clearer from context that the rent-return is the responsibility of the caller.

bartonjs commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -392,5 +411,99 @@ public int GetArrayRank()
+        }
+
+        /// <summary>
+        /// Returns a <see cref="TypeName" /> object representing a one-dimensional array
```
Methods always return (well, except ones that [DoesNotReturn]...).  I think the verb you want here is "Creates", like in the previous method.

This applies to the remainder of the Make- methods.

adamsitnik committed:
address code review feedback

adamsitnik committed:
rename MakeSimpleTypeName to WithAssemblyName

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -392,5 +411,100 @@ public int GetArrayRank()
+        {
+            if (!IsSimple)
+            {
+                TypeNameParserHelpers.ThrowInvalidOperation_NotSimpleName(FullName);
```
`FullName` can produce non-linear amount of work in case a long assembly name was applied to multiple generic arguments. Is it a concern? Should we omit the `FullName` in the exception message?

adamsitnik replied:
The NrbfDecoder won't ever call this method for non-simple types, so I believe it's acceptable.

jkotas replied:
I do not have a problem with it. 

I think you need to be careful in the threat-modeling about any place (including logging or error handling) that gets the full name since it may produce non-linear amount of work. Reducing the number of places that get the full name can make this easier. It is sort of defense-in-depth.

bartonjs replied:
Exception messages, particularly for this library, are an interesting beast.

The only writer I know of for NRBF is BF (and the remote procedure call parts of remoting, but it doesn't save answers).  BF won't mangle type names (more than it already does), and won't write down unsupported records (or we would have supported them).  So, where is the person getting this bad data other than from corrupted/tampered/malicious data?  They probably don't care what kind of record isn't supported, it just means the data's bad.  They probably don't care what the invalid type name was, it just means the data's bad.

Someone trying to write a new NRBF writer might end up caring, so we shouldn't discard the data, but it could easily be stored in either the property bag or a named property on a new derived exception type (if we felt it needed it), and the 0.01% who actually want the data can get it, while the 99.99% who don't care don't ever have to worry about the safety of logging exception messages.

A nicer middle ground is to capture numeric/enummy things in the message, but leave things like "arbitrary text" out.

(OK, this comment ended up lined up in S.R.MD.TypeName when I thought it was in NRBF... but pretend it's in NRBF)

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -392,5 +411,100 @@ public int GetArrayRank()
+        {
+            if (!IsSimple)
+            {
+                TypeNameParserHelpers.ThrowInvalidOperation_NotSimpleName(FullName);
```
Same here

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -392,5 +411,100 @@ public int GetArrayRank()
+        public TypeName MakeArrayTypeName(int rank)
+            => rank <= 0
+                ? throw new ArgumentOutOfRangeException(nameof(rank))
+                : MakeElementTypeName((sbyte)rank);
```
`(sbyte)` cast can overflow here

adamsitnik replied:
Another great catch ;)

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -69,6 +69,25 @@ sealed class TypeName
+#if SYSTEM_REFLECTION_METADATA
+        private TypeName(string? fullName,
+            AssemblyNameInfo? assemblyName,
+            TypeName? elementOrGenericType = default,
+            TypeName? declaringType = default,
+            ImmutableArray<TypeName> genericTypeArguments = default,
```
```suggestion
            TypeName? elementOrGenericType,
            TypeName? declaringType,
            ImmutableArray<TypeName> genericTypeArguments,
```
These default values do not seem to be useful. The code specifies them explicitly in nearly all cases.

jkotas reviewed (approved):
System.Reflection.Metadata part LGTM modulo comments. Thank you!

bartonjs commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/TypeNameHelpers.cs:
```diff
@@ -0,0 +1,183 @@
+
+    internal static PrimitiveType GetPrimitiveType<T>()
+    {
+        if (typeof(T) == typeof(bool)) return PrimitiveType.Boolean;
+        else if (typeof(T) == typeof(byte)) return PrimitiveType.Byte;
+        else if (typeof(T) == typeof(sbyte)) return PrimitiveType.SByte;
+        else if (typeof(T) == typeof(char)) return PrimitiveType.Char;
+        else if (typeof(T) == typeof(short)) return PrimitiveType.Int16;
+        else if (typeof(T) == typeof(ushort)) return PrimitiveType.UInt16;
+        else if (typeof(T) == typeof(int)) return PrimitiveType.Int32;
+        else if (typeof(T) == typeof(uint)) return PrimitiveType.UInt32;
+        else if (typeof(T) == typeof(long)) return PrimitiveType.Int64;
+        else if (typeof(T) == typeof(ulong)) return PrimitiveType.UInt64;
+        else if (typeof(T) == typeof(float)) return PrimitiveType.Single;
+        else if (typeof(T) == typeof(double)) return PrimitiveType.Double;
+        else if (typeof(T) == typeof(decimal)) return PrimitiveType.Decimal;
+        else if (typeof(T) == typeof(DateTime)) return PrimitiveType.DateTime;
+        else if (typeof(T) == typeof(TimeSpan)) return PrimitiveType.TimeSpan;
+        else if (typeof(T) == typeof(string)) return PrimitiveType.String;
+        else if (typeof(T) == typeof(IntPtr)) return IntPtrPrimitiveType;
+        else if (typeof(T) == typeof(UIntPtr)) return UIntPtrPrimitiveType;
+        else throw new InvalidOperationException();
```
Braces aren't required.  Newlines are.

```suggestion
        if (typeof(T) == typeof(bool))
            return PrimitiveType.Boolean;
        else if (typeof(T) == typeof(byte))
            return PrimitiveType.Byte;
        else if (typeof(T) == typeof(sbyte))
            return PrimitiveType.SByte;
        else if (typeof(T) == typeof(char))
            return PrimitiveType.Char;
        else if (typeof(T) == typeof(short))
            return PrimitiveType.Int16;
        else if (typeof(T) == typeof(ushort))
            return PrimitiveType.UInt16;
        else if (typeof(T) == typeof(int))
            return PrimitiveType.Int32;
        else if (typeof(T) == typeof(uint))
            return PrimitiveType.UInt32;
        else if (typeof(T) == typeof(long))
            return PrimitiveType.Int64;
        else if (typeof(T) == typeof(ulong))
            return PrimitiveType.UInt64;
        else if (typeof(T) == typeof(float))
            return PrimitiveType.Single;
        else if (typeof(T) == typeof(double))
            return PrimitiveType.Double;
        else if (typeof(T) == typeof(decimal))
            return PrimitiveType.Decimal;
        else if (typeof(T) == typeof(DateTime))
            return PrimitiveType.DateTime;
        else if (typeof(T) == typeof(TimeSpan))
            return PrimitiveType.TimeSpan;
        else if (typeof(T) == typeof(string))
            return PrimitiveType.String;
        else if (typeof(T) == typeof(IntPtr))
            return IntPtrPrimitiveType;
        else if (typeof(T) == typeof(UIntPtr))
            return UIntPtrPrimitiveType;
        else
            throw new InvalidOperationException();
```

bartonjs reviewed (dismissed):
Style violations must be corrected.

adamsitnik committed:
Apply suggestions from code review

Co-authored-by: Jeremy Barton <jbarton@microsoft.com>
Co-authored-by: Jan Kotas <jkotas@microsoft.com>

adamsitnik committed:
don't use sbyte to store array rank, the limit of 32 can be changed in the future and should not be enforced by the parser

adamsitnik committed:
fix the build
